cmd/service-controller/controller.go

```diff
@@ -309,15 +320,24 @@ func (c *Controller) checkServiceFor(desired *ServiceBindings, actual *corev1.Se
 	if len(actual.Spec.Ports) > 0 {
 		if actual.Spec.Ports[0].Port != int32(desired.publicPort) {
 			update = true
+			if !hasOriginalPort(*actual) {
+				actual.ObjectMeta.Annotations[types.OriginalPortQualifier] = strconv.FormatInt(int64(actual.Spec.Ports[0].Port), 10)
+			}
 			actual.Spec.Ports[0].Port = int32(desired.publicPort)
 		}
 		if actual.Spec.Ports[0].TargetPort.IntValue() != desired.ingressPort {
 			update = true
+			if !hasOriginalTargetPort(*actual) {
```

Comment:
Should we always update the originalTargetPort when modifying the current value? E.g. service is created and annotated, skupper modifies it including an update to the target port, user now updates the service themselves including a change to the target port, skupper will again update it, but now if restored it would set it back to the very first port.

It does seem simpler to reason about perhaps to always just save the last version we overwrote?

Reply:
That is a good point. If user modifies the target port, so that the OriginalTargetPort annotation
exists but the "new" target port is different than the current one, it makes sense to me to update
the value of the OriginalTargetPort annotation.

Reply:
I believe that won't work. In example, if we change the protocol, the `desired.ingressPort`
will change and will be different from `actual.Spec.Ports[0].TargetPort`, so updating the
`originalTargetPort` will make it unrecoverable.
Maybe if that is not too much, we can check if protocol has not changed, then in this case
I believe it makes sense to update the `originalTargetPort`. What do you think?
not changed?

Reply:
Stepping back, I think the root of the problem - for the case where target is not specified and therefore where we are editing the service that we are also using as the input into the definition - is that we need to be able to clearly distinguish changes made by the user from those made by the controller.

So perhaps what we could do is have two annotations per field of conflict. E.g. in this case originalTargetPort and assignedTargetPort. When the controller edits the TargetPort on the service, it first checks whether the current value matches the current assignedTargetPort annotation. If it does not, then it updates the originalTargetPort. Either way it then updates the service TargetPort and the assignedTargetPort.

The definition monitor can then also ignore the TargetPort if it matches the assignedTargetPort.

We could use a similar scheme for the selector. Does that make sense? Does it sound easy to understand and reason about?

Reply:
Lets say the service is annotated with `skupper.io/proxy=tcp` only, with `spec.selector.targetPort=8080`.

Skupper will update it and, lets say, update target port to `1024`. When doing so, it will set
`originalTargetPort` to 8080 and `assignedTargetPort to `1024`.

Then user switches the protocol to `http`, controller will update the assigned target port to (i.e.) `1025`
and then it will differ from the `assignedTargetPort`.

Maybe we should store the `originalProtocol` or `originalProxy` instead and update the target port when
it has been modified? If the protocol has not changed, but the target port is now different, then we should
update the `originalTargetPort` and update it (as this would mean a user change and not a controller change).

Does it make sense?

Reply:
Oh, so you mean comparing the *actual* target port (defined in the service) with the `assignedTargetPort` annotation? Correct, it makes sense. Sorry, but I guess I did not get it right earlier.
It seems correct to me and I will change/test it.

Reply:
Done. It seems to be working just fine now.